File: backend/app/schemas/audit.py
"""감사 로그 조회 API 스키마 (T-32)."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from app.core.timezone import local_isoformat


class AuditLogResponse(BaseModel):
    """감사 로그 1건 응답."""

    id: int
    actor_user_id: int | None = None
    actor_label: str | None = None
    action: str
    resource_type: str | None = None
    resource_id: str | None = None
    result: str
    occurred_at_utc: datetime
    # 화면 표시용 로컬(APP_TIMEZONE) 시각 문자열. occurred_at_utc는 정렬/필터 기준으로 유지.
    occurred_at_local: str
    ip_address: str | None = None
    meta: dict | None = None

    @classmethod
    def from_model(cls, row) -> "AuditLogResponse":
        """AuditLog ORM 행을 응답으로 변환(occurred_at_local 파생 포함)."""
        return cls(
            id=row.id,
            actor_user_id=row.actor_user_id,
            actor_label=row.actor_label,
            action=row.action,
            resource_type=row.resource_type,
            resource_id=row.resource_id,
            result=row.result,
            occurred_at_utc=row.occurred_at_utc,
            occurred_at_local=local_isoformat(row.occurred_at_utc),
            ip_address=row.ip_address,
            meta=row.meta,
        )

File: frontend/src/types/audit.ts
/** 감사 로그(시스템 사용자 활동 이력) 타입 (백엔드 schemas/audit.py 대응). */

/** 감사 로그에 실제 기록되는 행위(action) 값. backend/app/core/constants.py AuditAction과 동기화. */
export type AuditAction =
  | 'login'
  | 'report_view'
  | 'report_create'
  | 'report_update'
  | 'report_delete'
  | 'report_visibility_change'
  | 'export_run'
  | 'mail_send'
  | 'mail_schedule_create'
  | 'mail_schedule_update'
  | 'mail_schedule_delete'
  | 'permission_change'
  | 'group_change'
  | 'refresh_trigger'
  | 'refresh_cancel'
  | 'collect_now'
  | 'admin_setting_change'
  | 'powerbi_api_failure'
  | 'permission_denied'
  | 'request_create'
  | 'request_update'
  | 'request_comment'
  | 'stats_view'

export type AuditResult = 'success' | 'failure'

export interface AuditLogItem {
  id: number
  actor_user_id: number | null
  actor_label: string | null
  action: string
  resource_type: string | null
  resource_id: string | null
  result: string
  occurred_at_utc: string
  /** APP_TIMEZONE(KST) 기준 표시용 로컬 시각 문자열. */
  occurred_at_local: string
  ip_address: string | null
  meta: Record<string, unknown> | null
}

/** GET /api/audit-logs 쿼리 파라미터. */
export interface AuditLogQuery {
  from?: string
  to?: string
  actorUserId?: number
  action?: string
  resourceType?: string
  result?: string
  q?: string
  limit?: number
  offset?: number
}
